### PR TITLE
feat(action): add action to select value based on condition

### PR DESCRIPTION
**Pull request by felix11h**
This composite action allows to conditionally pass a value to a subsequent step in a GitHub Action workflow. We will use this our `merge-checks.yml` for public repositories:

```yml
name: Merge checks

on:
  push:
    branches:
      - feature/*
  pull_request_target:

jobs:
  merge-checks:
    runs-on: ubuntu-22.04
    if: github.event_name == 'push' || github.event.pull_request.head.repo.full_name != github.repository
    permissions:
      contents: read
      statuses: write
    steps:
      - name: Get head commit hash
        id: get_head_hash
        uses: moneymeets/moneymeets-composite-actions/select-conditionally@master
        with:
            condition: ${{ github.event_name == 'pull_request_target' }}
            if_condition_true: ${{ github.event.pull_request.head.sha }}
            if_condition_false: ${{ github.sha }}

      - name: Merge checks
        uses: moneymeets/action-merge-checks@master
        with:
          github_token: ${{ secrets.GITHUB_TOKEN }}
          github_head_sha: ${{ steps.get_head_hash.outputs.value }}
```

### CODE REVIEW

File: select-conditionally/action.yml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    description: 'Selected value based condition'
+
+runs:
+  using: "composite"
+  steps:
+    - run: |
+        if [ "${{ inputs.condition }}" = "true" ]; then 
+          SELECTED_VALUE=${{ inputs.if_condition_true}} 
+        elif [ "${{ inputs.condition }}" = "false" ]; then 
+          SELECTED_VALUE=${{ inputs.if_condition_false }}
```
Comment (zyv) [inline select-conditionally/action.yml]:
```suggestion
          SELECTED_VALUE="${{ inputs.if_condition_true }}"
        elif [ "${{ inputs.condition }}" = "false" ]; then 
          SELECTED_VALUE="${{ inputs.if_condition_false }}"
```

I think quoting is important here in the case that values happen to contain spaces.

File: select-conditionally/action.yml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        else
+          echo "Expected value for condition is 'true' or 'false'. Received '${{ inputs.condition }}'" 
+          exit 1
+        fi
+        echo ::set-output name=value::$SELECTED_VALUE
```
Comment (zyv) [inline select-conditionally/action.yml]:
```suggestion
        echo "::set-output name=value::$SELECTED_VALUE"
```
Also here.

File: select-conditionally/action.yml [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+description: 'Select value based on condition'
+
+inputs:
+  condition:
+    description: 'Condition for value selection. Possible values: [true, false]'
+    required: true
+  if_condition_true:
+    description: 'Value to return if condition true'
+    required: true
+  if_condition_false:
```
Comment (zyv) [inline select-conditionally/action.yml]:
```suggestion
  if_true:
    description: 'Value to return if condition true'
    required: true
  if_false:
```

It's a matter of taste, but I would have shortened this as above.

File: select-conditionally/action.yml [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+runs:
+  using: "composite"
+  steps:
+    - run: |
+        if [ "${{ inputs.condition }}" = "true" ]; then 
+          SELECTED_VALUE=${{ inputs.if_condition_true}} 
+        elif [ "${{ inputs.condition }}" = "false" ]; then 
+          SELECTED_VALUE=${{ inputs.if_condition_false }}
+        else
+          echo "Expected value for condition is 'true' or 'false'. Received '${{ inputs.condition }}'"
```
Comment (zyv) [inline select-conditionally/action.yml]:
```suggestion
          echo "Expected value for condition is 'true' or 'false'. Received '${{ inputs.condition }}'." 
```

**Review by zyv — CHANGES_REQUESTED** [review]
I don't know if you feel strongly about the name of the action, but I would have rather called it `conditional-expression`. Sad that GitHub doesn't offer conditional expressions in string interpolation.

I also was about to ask whether you think it even makes sense to have it as an action instead of packing bash in the run value, but after I tried it, it looked somehow ugly... and it's easy to get things wrong with quoting, etc.

So probably an action like you decided to do is the best solution, even though at first it looks like an overkill.

**Comment by felix11h** [comment]
Thank you for checking this action carefully, @zyv! I will change the name, no problem at all, and still clean up after I have added your suggestions directly just now.

**Review by mm-ivanovig — APPROVED** [review]

File: conditional-expression/action.yml [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    - run: |
+        if [ "${{ inputs.condition }}" = "true" ]; then 
+          SELECTED_VALUE="${{ inputs.if_true }}"
+        elif [ "${{ inputs.condition }}" = "false" ]; then 
+          SELECTED_VALUE="${{ inputs.if_false }}"
+        else
+          echo "Expected value for condition is 'true' or 'false'. Received '${{ inputs.condition }}'." 
+          exit 1
+        fi
+        echo "::set-output name=value::$SELECTED_VALUE"
```
Comment (zyv) [inline conditional-expression/action.yml:31]:
```suggestion
        echo "::set-output name=value::'$SELECTED_VALUE'"
```
I think I also missed this part...

**Review by zyv — APPROVED** [review]
I think one more quoting issue I overlooked...

By the way, it occurred to me, that you could have dome something like this with the official `github-script` action which I have been advertising several times already instead of using Bash and struggling with the arcane syntax and quoting issues:

```yaml
      - uses: actions/github-script@v6
        id: select_value
        with:
          script: |
            core.setOutput("value", core.getBooleanInput('condition') ? core.getInput('if_true') : core.getInput('if_false'));  
```

Oh wait, but then again, do we need a composite action or we can do it directly in Pulumi? :)


```yaml
      - uses: actions/github-script@v6
        id: get-head-hash  # Note that GitHub convention is to always use dashes in step ids
        with:
          result-encoding: string
          script: |
            core.setOutput("value", context.eventName == "pull_request_target" ? context.payload.pull_request.head.sha : context.sha);

        # Result available as ${{ steps.get-head-hash.outputs.result }}
```

How about that?

**Comment by felix11h** [comment]
> How about that?

Oh, this looks great, thank you. I'm testing this right now.

**Comment by felix11h** [comment]
Your suggestion works, @zyv. I have updated https://github.com/moneymeets/moneymeets-pulumi/pull/291, and I'm closing this pull request. Thank you!

**Pull request closed without merging**